feat: Add support for disaggregation with pp with pytorch backend

pcastonguay opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enhanced integration tests for disaggregated model serving with expanded support for various tensor and pipeline parallelism configurations.
  * Added new YAML configuration files to enable flexible context and generation server setups.
  * Introduced dynamic retrieval of accuracy task classes by name in integration tests.

* **Bug Fixes**
  * Improved error handling when removing files during build processes.

* **Tests**
  * Added comprehensive parameterized tests covering symmetric and asymmetric parallelism scenarios for disaggregated serving.
  * Extended test coverage with new test functions and configurations for TinyLlama-1.1B-Chat-v1.0.

* **Chores**
  * Improved logging for configuration mismatches and KV cache handling in pipeline parallel execution.

* **Refactor**
  * Internal enhancements to resource management for key-value cache parallelism tracking.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

[JIRA ticket/NVBugs ID/GitHub issue][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR to support a new feature about cache manager for JIRA ticket TRTLLM-1000, it would be like:

[TRTLLM-1000][feat] Support a new feature about cache manager

Or I have a PR to fix a Llama3 accuracy issue:

[https://nvbugs/1234567][fix] Fix Llama3 accuracy issue
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

pcastonguay committed:
ctx_pp2_gen_pp1_draft

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay committed:
Fix issue with nb_kv_heads_per_layer

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay committed:
Allowing no inflight reqs with disagg

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

raayandhar committed:
adding some initial tests for disagg+pp

Signed-off-by: raayandhar <rdhar@nvidia.com>

raayandhar committed:
add asymmetric disagg PP tests

Signed-off-by: raayandhar <rdhar@nvidia.com>

raayandhar committed:
fix a test and small changes for disagg+pp

Signed-off-by: raayandhar <rdhar@nvidia.com>

reasonsolo committed:
fix disagg+pp tests

Signed-off-by: Lizhi Zhou <1432185+reasonsolo@users.noreply.github.com>

pcastonguay committed:
Minor fixes

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #13220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13220/) [ run ] triggered by Bot

Tabrizian commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -155,18 +155,33 @@ def __init__(
                 (num_kv_heads + tp_size - 1) // tp_size
                 for _ in range(self.num_local_layers)
             ]
+            self.total_num_kv_heads_per_layer = [
```
What is the difference between `total_num_kv_heads_per_layer` and `num_kv_heads_per_layer`? Is one of them for the layers that each partition is responsible for.

Tabrizian commented on tests/integration/defs/disaggregated/test_disaggregated.py:
```diff
@@ -540,6 +550,106 @@ def test_disaggregated_ngram(disaggregated_test_root, llm_venv,
+@pytest.mark.skip_less_device(4)
+@pytest.mark.parametrize("llama_model_root", ['TinyLlama-1.1B-Chat-v1.0'],
+                         indirect=True)
+def test_disaggregated_ctxpp2_genpp2(disaggregated_test_root, llm_venv,
```
The tests need to be registered.

Tabrizian commented on tests/integration/defs/accuracy/test_disaggregated_serving.py:
```diff
@@ -315,6 +334,69 @@ def test_eagle3(self, overlap_scheduler):
+        return self.run_parallel_test(pp, tp, pp, tp,
+                                      get_accuracy_task(testset))
+
+    # We focus on ctx+pp and gen+tp usecases for RTX6000D
```
Should we remove this comment?

Tabrizian commented on tests/integration/defs/accuracy/test_disaggregated_serving.py:
```diff
@@ -315,6 +334,69 @@ def test_eagle3(self, overlap_scheduler):
+    def run_parallel_test(self, ctx_pp: int, ctx_tp: int, gen_pp: int,
+                          gen_tp: int, test_set: LlmapiAccuracyTestHarness):
+        if ctx_tp * ctx_pp + gen_tp * gen_pp > get_device_count():
+            pytest.skip(
```
```suggestion
            pytest.fail(
```

raayandhar replied:
What's the reason for failing instead of skipping?

pcastonguay commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -643,6 +644,7 @@ def _need_return_log_probs(self, scheduled_requests: ScheduledRequests):
         return False
 
     def _executor_loop_pp(self):
+        logger.info(f"Starting executor loop for pp_rank {self.dist.pp_rank}")
```
Yes I think so.

tensorrt-cicd commented:
[PR_Github #13235](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13235/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9882](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9882/) completed with status: 'FAILURE'

raayandhar committed:
minor fix + register tests

Signed-off-by: raayandhar <rdhar@nvidia.com>

pcastonguay commented:
> I have a commit on my [branch](https://github.com/NVIDIA/TensorRT-LLM/commit/5a2c2ae1aa4d1a7cb50387d65d2568122d4c227c) that addresses the feedback. Despite accepting the invitation to @pcastonguay's branch, there may some permission issue or otherwise because I cannot push upstream to his branch. Perhaps the best is for @pcastonguay to cherry-pick my commit? I am still verifying the tests, they are taking a very long time on my side, could also be the ongoing node issues. @reasonsolo did you test all of them?

I cherry-picked your changes

pcastonguay commented:
> I have a commit on my [branch](https://github.com/NVIDIA/TensorRT-LLM/commit/5a2c2ae1aa4d1a7cb50387d65d2568122d4c227c) that addresses the feedback. Despite accepting the invitation to @pcastonguay's branch, there may some permission issue or otherwise because I cannot push upstream to his branch. Perhaps the best is for @pcastonguay to cherry-pick my commit? I am still verifying the tests, they are taking a very long time on my side, could also be the ongoing node issues. @reasonsolo did you test all of them?

I cherry-picked your changes

pcastonguay commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13255](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13255/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #9899](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9899/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #13258](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13258/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13258](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13258/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #9901](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9901/) completed with status: 'FAILURE'

pcastonguay committed:
Fixing disagg test

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay commented on tests/integration/defs/accuracy/test_disaggregated_serving.py:
```diff
@@ -315,6 +334,68 @@ def test_eagle3(self, overlap_scheduler):
+    def run_parallel_test(self, ctx_pp: int, ctx_tp: int, gen_pp: int,
+                          gen_tp: int, test_set: LlmapiAccuracyTestHarness):
+        if ctx_tp * ctx_pp + gen_tp * gen_pp > get_device_count():
+            pytest.fail(
```
Fixed.

raayandhar commented on tests/integration/defs/disaggregated/test_disaggregated.py:
```diff
@@ -540,6 +550,106 @@ def test_disaggregated_ngram(disaggregated_test_root, llm_venv,
                            cwd=llm_venv.get_working_directory())
 
 
+@pytest.mark.skip_less_device(4)
```
I made a small typo here: this should be 8 and not 4

pcastonguay replied:
Fixed.

tensorrt-cicd commented:
[PR_Github #13266](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13266/) [ run ] triggered by Bot

raayandhar commented on tests/integration/defs/disaggregated/test_disaggregated.py:
```diff
@@ -540,6 +550,106 @@ def test_disaggregated_ngram(disaggregated_test_root, llm_venv,
+                           cwd=llm_venv.get_working_directory())
+
+
+@pytest.mark.skip_less_device(4)
```
I made a small typo here: this should be 8 and not 4

pcastonguay committed:
More fixes to tests

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #13272](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13272/) [ run ] triggered by Bot

raayandhar commented:
/bot run --disable-fail-fast

raayandhar commented:
The ctx pp4 + gen pp4 integration test doesn't work right now, it just times out with some useless error message. We should make sure not to merge while the test is broken. Just as a note.

reasonsolo committed:
fix tp_size error in launch_diaggregated_llm

Signed-off-by: Lizhi Zhou <1432185+reasonsolo@users.noreply.github.com>

pcastonguay committed:
Disabling 8 GPU tests

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay committed:
Fixing skip gpu less than

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #13369](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13369/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #13372](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13372/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `1a86b15`

raayandhar commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #13396](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13396/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13396](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13396/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #13415](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13415/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10034 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10034/) completed with status: 'FAILURE'

raayandhar commented:
/bot run --stage-list "GB200-4_GPUs-PyTorch-1"

tensorrt-cicd commented:
[PR_Github #13430](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13430/) [ run ] triggered by Bot

pcastonguay commented:
/bot run --only-multi-gpu-test --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #13430](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13430/) [ run ] completed with state `ABORTED`

pcastonguay committed:
Allowing empty scheduled batches for pp

Signed-off-by: Patrice Castonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay commented:
/bot run --disable-fail-fast --add-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #13437](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13437/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13432](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13432/) [ run ] completed with state `ABORTED`

pcastonguay committed:
Update tests/integration/defs/accuracy/test_disaggregated_serving.py

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: pcastonguay <55748270+pcastonguay@users.noreply.github.com>

pcastonguay commented on tests/integration/test_lists/test-db/l0_dgx_h200.yml:
```diff
@@ -23,6 +23,14 @@ l0_dgx_h200:
+  - accuracy/test_disaggregated_serving.py::TestLlama3_1_8BInstruct::test_ctx_pp_gen_tp_asymmetric[MMLU-gen_tp=1-ctx_pp=4]
+  - accuracy/test_disaggregated_serving.py::TestLlama3_1_8BInstruct::test_ctx_pp_gen_tp_asymmetric[MMLU-gen_tp=2-ctx_pp=4]
+  - disaggregated/test_disaggregated.py::test_disaggregated_ctxtp2pp2_gentp2pp2[TinyLlama-1.1B-Chat-v1.0]
+  - disaggregated/test_disaggregated.py::test_disaggregated_ctxpp4_genpp4[TinyLlama-1.1B-Chat-v1.0]
```
@raayandhar won't that test fail then?

pcastonguay commented:
/bot run --stage-list "H100_PCIe-CPP-1, H100_PCIe-TensorRT-Perf-1"

tensorrt-cicd commented:
[PR_Github #13456](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13456/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13437](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13437/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #13456](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13456/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10073 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10073/) completed with status: 'FAILURE'

reasonsolo committed:
Merge branch 'main' into pyt_disagg_pp

tensorrt-cicd commented:
[PR_Github #13479](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13479/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13479](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13479/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10093 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10093/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #13505](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13505/) [ run ] triggered by Bot

pcastonguay commented:
/bot skip --comment "All tests passed after running manually"

tensorrt-cicd commented:
[PR_Github #13541](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13541/) [ skip ] completed with state `SUCCESS`
Skipping testing for commit `e7dbdd3`
